src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,376 @@
+
+                var deleteCount = 5 + 5 * (updatedOrderedEditNumber - orderedEditNumber);
+                for (var i = 1; i <= updatedOrderedEditNumber - orderedEditNumber; i++)
+                {
+                    newGroupedSemanticTokens[editIndices[orderedEditNumber + i]].AddToEnd(tokensToInsert);
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: editStartPosition, deleteCount: deleteCount, data: tokensToInsert.ToArray()));
+                return updatedOrderedEditNumber;
+            }
+
+            static int AddInsertionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int orderedEditNumber,
+                SemanticToken groupedSemanticToken,
+                int editStartPosition)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                groupedSemanticToken.AddToEnd(tokensToInsert);
+
+                // An insert can only be combined with other inserts that directly follow it.
+                var updatedOrderedEditNumber = GetUpdatedOrderedEditNumber(
+                    SemanticTokenEditKind.Insert, indexToEditKinds, editIndices, orderedEditNumber);
+
+                for (var i = 1; i <= updatedOrderedEditNumber - orderedEditNumber; i++)
+                {
+                    newGroupedSemanticTokens[editIndices[orderedEditNumber + i]].AddToEnd(tokensToInsert);
+                }
+
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: editStartPosition, deleteCount: 0, data: tokensToInsert.ToArray()));
+                return updatedOrderedEditNumber;
+            }
+
+            static int AddDeletionEdit(
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int orderedEditNumber,
+                int editStartPosition)
+            {
+                // A deletion can only be combined with other deletions that directly follow it.
+                var updatedOrderedEditNumber = GetUpdatedOrderedEditNumber(
+                    SemanticTokenEditKind.Delete, indexToEditKinds, editIndices, orderedEditNumber);
+
+                var deleteCount = 5 + 5 * (updatedOrderedEditNumber - orderedEditNumber);
+                semanticTokensEdits.Add(
+                    GenerateEdit(start: editStartPosition, deleteCount: deleteCount, data: Array.Empty<int>()));
+                return updatedOrderedEditNumber;
+            }
+
+            // Returns the updated ordered edit number after we know how many edits we can combine.
+            static int GetUpdatedOrderedEditNumber(
```

Comment:
GetEndIndexOfSameEditKind?  (not sure if that's better).  but i do like saying "End"